File: sdk-js/src/policy.ts
// The policy engine — the firewall's brain (mirror of voan/policy.py).
// First matching rule wins; nothing matches => default-allow. Flip `default` to
// "block" for a deny-by-default posture.
import type { Action, DecisionType, Verdict } from "./schema.ts";
import { DEFAULT_RULES, ruleMatches, type Rule } from "./rules.ts";

export class PolicyEngine {
  rules: Rule[];
  default: DecisionType;

  constructor(rules?: Rule[], def: DecisionType = "allow") {
    this.rules = rules ? [...rules] : [...DEFAULT_RULES];
    this.default = def;
  }

  addRule(rule: Rule, front = true): this {
    if (front) this.rules.unshift(rule);
    else this.rules.push(rule);
    return this;
  }

  evaluate(action: Action): Verdict {
    for (const rule of this.rules) {
      if (ruleMatches(rule, action)) {
        return {
          decision: rule.decision, rule: rule.id, code: rule.code,
          severity: rule.severity, reason: rule.reason,
        };
      }
    }
    return {
      decision: this.default, rule: "default", code: "",
      severity: "Low", reason: "no rule matched",
    };
  }
}

File: voan/policy.py
"""The policy engine — the firewall's brain.

Given an Action, it returns a Verdict. First matching rule wins; if nothing
matches the action is allowed (default-allow, like a default-permit firewall —
flip `default` to BLOCK for a deny-by-default posture). Rules are ordered, so
put the most dangerous/most specific signatures first in rules.py.
"""
from .rules import DEFAULT_RULES, Rule
from .schema import Action, Decision, Verdict


class PolicyEngine:
    def __init__(self, rules=None, default=Decision.ALLOW):
        # Copy so callers can mutate their engine without touching the defaults.
        self.rules = list(rules if rules is not None else DEFAULT_RULES)
        self.default = default

    def add_rule(self, rule: Rule, front=True):
        """Register a custom rule. Front-loaded by default so user policy can
        override a built-in signature."""
        self.rules.insert(0, rule) if front else self.rules.append(rule)
        return self

    def evaluate(self, action: Action) -> Verdict:
        for rule in self.rules:
            if rule.matches(action):
                return Verdict(decision=rule.decision, rule=rule.id,
                               code=rule.code, severity=rule.severity,
                               reason=rule.reason)
        return Verdict(decision=self.default, rule="default",
                       reason="no rule matched")
